### ISSUE TITLE
Retry when DynamoDB source fails to initialize due to permission issue

### ISSUE THREAD

**Issue by oeyh** [opensearch-project/data-prepper#3623]
**Is your feature request related to a problem? Please describe.**
Currently, if DDB source role does not have `dynamodb:DescribeTable` permission, the pipeline will shut down during initialization. We can catch this as a retryable error like we did for Opensearch sink to allow user to correct permissions on the role.

The permission error is like this:
```
[dynamodb-pipeline-sink-worker-2-thread-1] ERROR org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor - Pipeline [dynamodb-pipeline] process worker encountered a fatal exception, cannot proceed further
java.util.concurrent.ExecutionException: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: User: arn:aws:sts::123456789012:assumed-role/pipelineRole/Data-Prepper-0233fade-6af1-4da2-a36b-923d9802c8cb is not authorized to perform: dynamodb:DescribeTable on resource: arn:aws:dynamodb:us-east-1:123456789012:table/test-source-table because no identity-based policy allows the dynamodb:DescribeTable action (Service: DynamoDb, Status Code: 400, Request ID: 839GJP2AL1PSOC8225BKVCULM3VV4KQNSO5AEMVJF66Q9ASUAAJG)
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor.afterExecute(PipelineThreadPoolExecutor.java:70) ~[data-prepper-core-2.6.0-SNAPSHOT.jar:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1129) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: User: arn:aws:sts::193777858833:assumed-role/pipelineRole/Data-Prepper-0233fade-6af1-4da2-a36b-923d9802c8cb is not authorized to perform: dynamodb:DescribeTable on resource: arn:aws:dynamodb:us-east-1:193777858833:table/gameday-test-source-table because no identity-based policy allows the dynamodb:DescribeTable action (Service: DynamoDb, Status Code: 400, Request ID: 839GJP2AL1PSOC8225BKVCULM3VV4KQNSO5AEMVJF66Q9ASUAAJG)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleErrorResponse(CombinedResponseHandler.java:125) ~[sdk-core-2.20.67.jar:?]
	...
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:56) ~[aws-core-2.20.67.jar:?]
	at software.amazon.awssdk.services.dynamodb.DefaultDynamoDbClient.describeTable(DefaultDynamoDbClient.java:2344) ~[dynamodb-2.20.67.jar:?]
	at org.opensearch.dataprepper.plugins.source.dynamodb.DynamoDBService.getTableInfo(DynamoDBService.java:271) ~[dynamodb-source-2.6.0-SNAPSHOT.jar:?]
...
```

**Describe the solution you'd like**

**Describe alternatives you've considered (Optional)**

**Additional context**

**Comment by daixba** [comment]
Thanks for the issue, this should be an easy fix. We can simply `give up` the init partition and so that the initialization process can repeat again and agin.

But a question behind this is that how does customer know there are something wrong with the pipeline, they should take action? Any thoughts?

**Comment by oeyh** [comment]
We log the error messages and when users notice that data is not being ingested into the sink and then check Data Prepper logs, they should be able to identify the issue and take actions.

**Comment by daixba** [comment]
Make senses, only DynamoDB is a little special. At the beginning, there may not have any data to be ingested into the sink as it's waiting for the export job to be completed, which can take a few minutes or even hours.

I will provide a fix as suggested to have a retry on this.

**Comment by daixba** [comment]
The fix in #3614 doesn't seem to resolve this issue.

Although it does retry on permission issue, but some how it keep retring even after permission are fixed. Will need to dive deeper to understand.

```2023-11-13T02:11:35.067 [pool-13-thread-1] INFO  org.opensearch.dataprepper.plugins.source.dynamodb.leader.LeaderScheduler - Try to initialize DynamoDB service
2023-11-13T02:11:35.109 [pool-13-thread-1] ERROR org.opensearch.dataprepper.plugins.source.dynamodb.leader.LeaderScheduler - Unable to call DescribeTableRequest to get information for table ... due to User: ... is not authorized to perform: dynamodb:DescribeTable on resource: ... because no identity-based policy allows the dynamodb:DescribeTable action (Service: DynamoDb, Status Code: 400, Request ID: ...)
2023-11-13T02:11:35.110 [pool-13-thread-1] ERROR org.opensearch.dataprepper.plugins.source.dynamodb.leader.LeaderScheduler - Exception occurred: Unable to get table information for .... Please make sure the permission is properly set
2023-11-13T02:12:35.110 [pool-13-thread-1] ERROR org.opensearch.dataprepper.plugins.source.dynamodb.leader.LeaderScheduler - Exception occurred: null
2023-11-13T02:13:35.111 [pool-13-thread-1] ERROR org.opensearch.dataprepper.plugins.source.dynamodb.leader.LeaderScheduler - Exception occurred: null```
